### DISCUSSION TITLE
M-ConFIG in TorchJD

### DISCUSSION THREAD

**Discussion by ValerianRey** [tum-pbs/ConFIG#19] [General]
Hi,
We recently added support for stateful aggregators in [TorchJD](https://github.com/SimplexLab/TorchJD) (a library where currently only the basic stateless version of ConFIG is implemented: https://torchjd.org/stable/docs/aggregation/config/).

Feel free to make a PR if you wanna add M-ConFIG to it. For reference, here are the pull requests related to ConFIG in TorchJD:
- Original PR proposing ConFIG: https://github.com/SimplexLab/TorchJD/pull/224/changes (closed because we didnt want the extra dependency)
- PR adding ConFIG: https://github.com/SimplexLab/TorchJD/pull/223 (merged)
- PR improving the implementation of ConFIG: https://github.com/SimplexLab/TorchJD/pull/281 (merged)

It's actually quite simple to add stateful aggregators now: all you have to do it inherit from the `torchjd.aggregation.Stateful` mixin, and implement a `reset` method. More details in https://github.com/SimplexLab/TorchJD/blob/main/CONTRIBUTING.md#adding-a-new-aggregator.

Note M-ConFIG is something that one of our users and recent contributors asked for recently (in our discord server).

**Comment by ValerianRey** [comment]
@qiauil (tagging you in case you didn't receive a notification)

**Comment by rkhosrowshahi** [comment]
Hi,
Thanks @ValerianRey for opening this discussion. 
We are maintaining [TorchJD](https://github.com/SimplexLab/TorchJD) to cover MOO methods. We haven't heard from ConFIG contributors regarding this discussion. So, I am kindly reminding and refreshing the discussion. 
We are looking forward to hearing back from contributors @qiauil and @thunil!

**Comment by qiauil** [comment]
Hi, @ValerianRey and @rkhosrowshahi,

Sorry for the late reply; I think the previous message wasn't notified in my email. Anyway, I think we have had a lot of discussions before on how to implement ConFIG and M-ConFIG in TorchJD, for example, https://github.com/SimplexLab/TorchJD/issues/131#issuecomment-2337218176. As far as I remember, M-ConFIG was not suitable for TorchJD at that moment, as M-ConFIG requires iterative calls to different loss functions in different training iterations, which is beyond the design scope of TorchJD. Not so sure about the current state of TorchJD, but I think I need to have a check.

Unfortunately, I haven't had too much time recently, so I will keep this discussion open and report progress here.

**Reply by ValerianRey** [reply]
My bad, I didn't look carefully enough at M-ConFIG before writing this. You're right that it doesn't just require a stateful aggregator.

I think the M-ConFIG algorithm, as it is written in your paper, could be decomposed as follows:
- First 4 lines of the for-loop can be done with a stateful object that maps a single gradient $\nabla_{\theta_{t-1}}\mathcal{L}_i$ to a Jacobian (the matrix of $\hat{m}$ in the algorithm).
- Line 5 is the ConFIG aggregator.
- Lines 6 to 10 are done in the Adam optimizer.

For the first point, we could one day have an object for that in a utils package of torchjd. But for now I think it's better if we focus on methods that can be implemented as aggregators, weightings, or even scalarizations in the near future. So let's postpone the integration of M-ConFIG to torchjd.

Sorry about the confusion!
